### PR TITLE
Add offline-to-online baselines for rlpd/ibrl/calql

### PR DESCRIPTION
**Pull request by jlidard**
This change is [Reviewable](https://reviewable.io/reviews/jlidard/dppo-dev/12)

### CODE REVIEW

File: model/diffusion/diffusion_ibrl.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+        # sample an action from the BC policy
+        bc_action = self.forward_sample(
+            cond=cond,
+            deterministic=True,
+            network_override=self.bc_policy,
+        )
+
+        # sample an action from the RL policy
+        rl_action = super().forward(
```
Comment (allenzren) [inline model/diffusion/diffusion_ibrl.py]: probably use self.forward_sample() instead? no need to use the parent class function?

File: model/diffusion/diffusion_ibrl.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+        # run all critics in batch
+        current_q = torch.vmap(self.critic_wrapper, in_dims=(0, 0, None))(
+            self.ensemble_params, self.ensemble_buffers, (obs, actions)
+        )  # (n_critics, B)
+        loss_critic = torch.mean((current_q - target_q[None]) ** 2)
+        return loss_critic
+
+    def loss_actor(self, obs):
+        action = self.forward_train(
```
Comment (allenzren) [inline model/diffusion/diffusion_ibrl.py]: self.forward()

File: model/diffusion/diffusion_ibrl.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            x = mean + std * noise
+
+            # clamp action at final step
+            if self.final_action_clip_value is not None and i == len(t_all) - 1:
+                x = torch.clamp(
+                    x, -self.final_action_clip_value, self.final_action_clip_value
+                )
+        return x
+
+    def forward_train(
```
Comment (allenzren) [inline model/diffusion/diffusion_ibrl.py]: remove

File: model/diffusion/diffusion_ibrl.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+            q_indices = torch.multinomial(q_weights, 1)
+            action = torch.where(
+                (q_indices == 0)[:, None],
+                bc_action,
+                rl_action,
+            )
+        return action
+
+    # override
+    @torch.no_grad()
```
Comment (allenzren) [inline model/diffusion/diffusion_ibrl.py]: remove, make it differentiable

**Review by allenzren — CHANGES_REQUESTED** [review]

File: model/diffusion/diffusion_ibrl.py (hunk 2 above)
Comment (jlidard) [reply model/diffusion/diffusion_ibrl.py]: Done.

File: model/diffusion/diffusion_ibrl.py (hunk 1 above)
Comment (jlidard) [reply model/diffusion/diffusion_ibrl.py]: Done.

File: model/diffusion/diffusion_ibrl.py (hunk 4 above)
Comment (jlidard) [reply model/diffusion/diffusion_ibrl.py]: Done.

File: model/diffusion/diffusion_ibrl.py (hunk 3 above)
Comment (jlidard) [reply model/diffusion/diffusion_ibrl.py]: Done.

**Review by jlidard — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/jlidard/dppo-dev/12)* status: 0 of 15 files reviewed, 4 unresolved discussions (waiting on @allenzren)

**Review by allenzren — APPROVED** [review]

**Comment by allenzren** [comment]
Do not merge.

**Pull request closed without merging**